soroban-sdk/src/cost_estimate.rs

```diff
@@ -0,0 +1,111 @@
+use soroban_env_host::{fees::FeeConfiguration, FeeEstimate, InvocationResources};
+
+use crate::{testutils::budget::Budget, Env};
+
+pub struct CostEstimate {
+    env: Env,
+}
+
+impl CostEstimate {
+    pub(crate) fn new(env: Env) -> Self {
+        Self { env }
+    }
+
+    /// Enables detailed per-invocation resource cost metering.
+    ///
+    /// The top-level contract invocations and lifecycle operations (such as
+    /// `register` or `env.deployer()` operations) will be metered and
+    /// all the metering information will reset in-between them. Metering will
+    /// not be reset while inside the call (e.g. if a contract calls or creates
+    /// another contract, that won't reset metering).
+    ///
+    /// The metered resources for the last invocation can be retrieved with
+    /// `resources()`, the estimated fee corresponding to these resources can be
+    /// retrieved with `fee()`, and the detailed CPU and memory memory
+    /// breakdown can be retrieved with `detailed_metering()`.
+    ///
+    /// While the resource metering may be useful for contract optimization,
+    /// keep in mind that resource and fee estimation may be imprecise. Use
+    /// simulation with RPC in order to get the exact resources for submitting
+    /// the transactions to the network.    
+    pub fn enable(&self) {
+        self.env.host().enable_invocation_metering();
+    }
+
+    /// Returns the resources metered during the last top level contract
+    /// invocation.    
+    ///
+    /// In order to get non-`None` results, `enable()` has to
+    /// be called and at least one invocation has to happen after that.
+    ///
+    /// Take the return value with a grain of salt. The returned resources mostly
+    /// correspond only to the operations that have happened during the host
+    /// invocation, i.e. this won't try to simulate the work that happens in
+    /// production scenarios (e.g. certain XDR rountrips). This also doesn't try
+    /// to model resources related to the transaction size.
+    ///
+    /// The returned value is as useful as the preceding setup, e.g. if a test
+    /// contract is used instead of a Wasm contract, all the costs related to
+    /// VM instantiation and execution, as well as Wasm reads/rent bumps will be
+    /// missed.    
+    pub fn resources(&self) -> Option<InvocationResources> {
```

Comment:
I think we should return the value, and panic on incorrect use.

Why: Convenience, and this is in tests, so the convenience of not having to unwrap something all the time is a big benefit. If someone makes a mistake it'll be in a test and they'll be able to fix it straight away. The panic can tell them what needs enabling exactly.

Reply:
Agree, updated. FWIW we only need to panic here, as `fee` just uses this.